java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java

```diff
@@ -0,0 +1,147 @@
+
+import java.text.MessageFormat;
+import java.util.List;
+import java.util.Map;
+
+@Rule(
+  key = "S2175",
+  name = "Inappropriate \"Collection\" calls should not be made",
+  tags = {"bug"},
+  priority = Priority.CRITICAL)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.DATA_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class CollectionInappropriateCallsCheck extends AbstractMethodDetection {
+
+  private final Map<String, String> primitiveToPrimitiveWrapper = buildPrimitiveMapper();
+
+  @Override
+  protected List<MethodInvocationMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      collectionMethodInvocation("remove"),
+      collectionMethodInvocation("contains")
+      );
+  }
+
+  private MethodInvocationMatcher collectionMethodInvocation(String methodName) {
+    return MethodInvocationMatcher.create()
+      .typeDefinition(TypeCriteria.subtypeOf("java.util.Collection"))
+      .name(methodName)
+      .addParameter("java.lang.Object");
+  }
+
+  @Override
+  protected void onMethodFound(MethodInvocationTree tree) {
+    Type argumentType = getType(tree.arguments().get(0));
+    Type collectionType = getMethodOwner(tree);
+    Type collectionParameterType = getTypeParameter(collectionType);
+
+    if (isKnown(collectionParameterType) && !isArgumentCompatible(argumentType, collectionParameterType)) {
+      addIssue(tree, MessageFormat.format("A \"{0}<{1}>\" cannot contain a \"{2}\"", collectionType, collectionParameterType, argumentType));
+    }
+  }
+
+  private Type getType(ExpressionTree tree) {
+    return ((AbstractTypedTree) tree).getSymbolType();
+  }
+
+  private Type getMethodOwner(MethodInvocationTree mit) {
+    if (mit.methodSelect().is(Kind.MEMBER_SELECT)) {
+      return getType(((MemberSelectExpressionTree) mit.methodSelect()).expression());
+    }
+    return ((MethodInvocationTreeImpl) mit).getSymbol().owner().getType();
+  }
+
+  @Nullable
+  private Type getTypeParameter(Type collectionType) {
+    if (collectionType instanceof ParametrizedTypeType) {
+      return getFirstTypeParameter((ParametrizedTypeType) collectionType);
+    }
+    return null;
+  }
+
+  @Nullable
+  private Type getFirstTypeParameter(ParametrizedTypeType parametrizedTypeType) {
+    for (TypeVariableType variableType : parametrizedTypeType.typeParameters()) {
+      return parametrizedTypeType.substitution(variableType);
+    }
+    return null;
+  }
+
+  private boolean isArgumentCompatible(Type argumentType, Type collectionParameterType) {
+    return (isSubtypeOf(argumentType, collectionParameterType) || isBoxingPossible(argumentType, collectionParameterType));
+  }
+
+  private boolean isSubtypeOf(Type argumentType, Type collectionParameterType) {
+    return argumentType.isSubtypeOf(collectionParameterType.getSymbol().getFullyQualifiedName());
+  }
+
+  private boolean isKnown(Type type) {
+    return type != null && !type.isTagged(Type.UNKNOWN);
```

Comment:
For raw types such as `List list = Lists.newArrayList();`, the test is required.